### PR TITLE
Tweak resample data code

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Internal changes to the ReSampleData class to better handle invalid input.

# Details

This is clean-up code which I've pulled out of https://github.com/sherpa/sherpa/pull/1735. PR #1777 relies on this PR.

In trying to add checks to improve coverage, I realised that the code - which was meant to report an error if ReSampleData was called with an invalid data object - did not actualy work, since it basically required there to be a y attribute, which is not guaranteed for non-1D data types. So, I've moved the check to earlier, taking advantage of the ndim field I recently-ish added (in #1477), rather than check on the data class (I added an earlier commit with a test of this functionality).
    
Should this class warn or error out if given a DataPHA or Data1DInt object? Alternatively, the code could be enhanced to properly handle this (but that is hard for the PHA case as we need to worry about the response).

### CODE REVIEW

**Comment by DougBurke** [comment]
@anetasie - just to say that this is one of the PRs I need attention.

**Comment by anetasie** [comment]
@DougBurke I will look at this one while traveling... nothing better to do on the plane.

**Comment by DougBurke** [comment]
I can think of several things I'd refer to do on an :aeroplane: ....

**Comment by DougBurke** [comment]
Rebased just to pick up recent chnages.

File: sherpa/sim/__init__.py [hunk 1]
```diff
@@ -828,6 +828,12 @@ class ReSampleData(NoNewAttributesAfterInit):
 
     """
     def __init__(self, data, model):
+
+        # Should this error out if data is an instance of Data1DInt?
+        if data.ndim != 1:
+            msg = f"{ReSampleData.__name__} {type(data)}"
```
Comment (hamogu) [inline sherpa/sim/__init__.py:834]: Should we be more wordy here, e.g. "Resample data does not implement support for Data2DInt". I know it's a NotImplementedError, I'm just trying to help a user understand what to do if they hit this error (namely, use a different data class).

**Comment by anetasie** [comment]
@DougBurke Is this resampling only used for the asymmetric errors? If that's the case then I don't think we should use DataPHA. On the other hand I can imagine that Data1DInt might be used.

**Comment by DougBurke** [comment]
@aneta - I didn't write this code and I wasn't involved in its development. All I can do is point to the existing code.

**Comment by anetasie** [comment]
@DougBurke Sure. If it is only used for the data with asymmetric errors then  the resampling is used to estimate the uncertainties with the implicit assumption of the Gaussian errors on data - different sigmas  for upper/lower error to account for assymetry.

**Comment by DougBurke** [comment]
I'm pretty sure the resample code does allow it to be run with symmetric
errors as well, but I haven't looked at this code for a while so I may be
missing something,

**Comment by anetasie** [comment]
Yes, it should be fine for the symmetric errors.

**Review by anetasie — APPROVED** [review]

File: sherpa/sim/__init__.py (hunk 1 above)
Comment (DougBurke) [reply sherpa/sim/__init__.py:834]: I don't have a lot of mental energy for improving the user experience here. If this call fails we don't have any replacement code the user could use - they'd have to do it themselves. I'm not sure how best to explain this without writing really-long error messages which then become too wordy. But as I said , I have little energy for this so if you can come up with a sensible message then I'm all ears.

Comment (hamogu) [reply sherpa/sim/__init__.py:834]:
```suggestion
            msg = f"{ReSampleData.__name__} is only implemented for 1D data, got {type(data)} instead."
```

**Comment by hamogu** [comment]
I did my suggestion inline, but did not change the test for that error. I don't have any other comments, so once you fix that test I'll approve.

**Review by hamogu — APPROVED** [review]
